rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,408 @@
+
+1. Once the lower store is "forked" in this way, there is no "merge".
+   The modified DB in the upper layer will completely cover up the lower store's DB.
+   If any new store objects are added to the lower store, the overlayFS-combined local store will never know.
+
+2. The database can be very large.
+   For example, Replit's 16 TB store has a 634 MB database.
+   OverlayFS doesn't know how to duplicate only part of a SQLite database, so we have to duplicate the whole thing per user/consumer.
+   This will waste lots of space with information the consumer may not care about.
+   And for a many-user product like Replit's, that will be wasting precious disk quota from the user's perspective too.
+
+## Use `nix-store --load-db` with the above
+
+As suggested [on discourse](https://discourse.nixos.org/t/super-colliding-nix-stores/28462/7), it is possible to augment the DB with additional paths after it has been created.
+Indeed, this is how the ISO installer for NixOS works.
+Running this periodically can allow the consumer to pick up any new paths added to the lower store since the last run.
+
+This does solve the first problem, but with poor performance penalty.
+We don't know what paths have changed, so we have slurp up the entire DB.
+With Replit's 16 TB store's DB, this took 10 minutes.
+
+We could out of band try to track "revisions" so just new paths are added, but then we are adding a new source of complexity vs the "statelessness" of on-demand accessing the lower store (and relying on monotonicity).
+
+## Bind mounts instead of overlayfs for `local-overlay` Store
+
+Instead of mounting the entire lower store dir underneath ours via OverlayFS, we could bind mount individual store objects as we need them.
+The bind-mounting store would not longer be the union of the lower store with the additional store objects, instead the lower store acts more as a special optimized substituter.
+
+This gives a normal form in that objects are bind-mounted if and only if they have a DB entry in the bind-mounting store;
+There is no "have store object, don't yet have DB entry" middle state to worry about.
+
+The downside of this is that Nix needs elevate permissions in order to create those bind mounts, and the impact of having arbitrarily many bind mounts is unknown.
+Even if this design works fine once set up, the imposition of an O(n) initialization setting up each bind mount is prohibitive for many use-cases.
+
+## Store implementations using FUSE
+
+We could have a single FUSE mount that could manually implement the "bind on demand" semantics described above without cluttering the mount namespace with an entry per each shared store object.
+FUSE however is quite primitive, in that every read need to be shuffled via the FUSE server.
+There is nothing like a "control plane vs data plane" separation where Nix could tell the OS "this directory is that other directory", and the OS can do the rest without involving the FUSE server.
```

Comment:
This is false, FUSE passthrough can lead to bypass the FUSE server.
There are multiple folks in the Linux filesystem ecosystem that are working on bringing FUSE ecosystems with a relatively on-par performance with in-kernel filesystem or reuse the existing filesystems.

https://lwn.net/Articles/932060/
https://github.com/extfuse/extfuse
https://lpc.events/event/16/contributions/1339/

I am not really convinced of not pursuing this alternative as I feel like this bring the maximum flexibility and compatibility for all the usecases instead of making it a very limited thing based on OverlayFS.

If you are interested into chatting more on how to make this alternative possible, feel free to ping me, I know quite about FUSE filesystems and I am planning to write a nixstorefs at some point, which will rely on FUSE semantics first.

I also challenge the "worse performance" than in-kernel mounting solutions, it would be good to bring data on that, if you only pay the `open` cost, this is quite cheap.

Reply:
Anyway, I have mixed feelings about the RFC because I feel like the FUSE approach is a much better route than this one, I can say that I am not satisfied and feel like it should be more ambitious if it is going to take a RFC route.

Reply:
Oh that is fantastic! That is getting at the heart of the problems with FUSE. I wish I had known about this earlier.

However, even if I did, I would have advocated starting with the OverlayFS approach. That is because all these things are not yet mainline, and kernel development / trying out bleeding edge features is vastly more expensive.

IMO the right thing to do is

1. Accept what we have on an experimental basis, using it to drum up interests and move us towards being able to pool resources on this.
2. Get in communication with these Kernel devs; indeed I was already emailing back and forth with Amir Goldstein about some restrictions in OverlayFS.
3. Try to be an early adopter of this stuff as it matures; nudge its development so it better meets our needs.

Also CC @flokli, because Tvix may be better positioned to be at the vanguard of trying this stuff out, as they are already exploring FUSE.

Reply:
Disclosure: I am also a tvix-store developer ;)

Reply:
Alright, your argument is sound and convinced me :).

Reply:
Oh my bad I didn't realize that. @RaitoBezarius you should shepherd this :).

Reply:
Aaaaaaaaaaa, OK for the nerdsnipe.

Reply:
There may be another approach with FUSE : using [MergerFS](https://github.com/trapexit/mergerfs)  with `symlinkify=true` and `cache.symlinks=true`. This way, FUSE is only used to resolve the symlink:
 - for a lot of applications, the opened directory fd will be reused, so most operations will run directly on the underlying store
 - later uses of the symlink would be cached by the kernel, so even without fuse-ebpf, we'd need the jump to userspace only once per symlink